engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java

```diff
@@ -0,0 +1,236 @@
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.state.migration.to_8_4.corrections;
+
+import static io.camunda.zeebe.db.impl.ZeebeDbConstants.ZB_DB_BYTE_ORDER;
+
+import io.camunda.zeebe.db.ColumnFamily;
+import io.camunda.zeebe.db.TransactionContext;
+import io.camunda.zeebe.db.ZeebeDb;
+import io.camunda.zeebe.db.impl.DbBytes;
+import io.camunda.zeebe.db.impl.DbCompositeKey;
+import io.camunda.zeebe.db.impl.DbInt;
+import io.camunda.zeebe.db.impl.DbLong;
+import io.camunda.zeebe.db.impl.DbString;
+import io.camunda.zeebe.engine.state.message.DbMessageState;
+import io.camunda.zeebe.engine.state.migration.DbMigratorImpl;
+import io.camunda.zeebe.protocol.ZbColumnFamilies;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used to correct the column family prefix for the DMN decision key by decision id
+ * and version which may contain entries for the MessageStats column family. Vice-versa correction
+ * is not needed because no data was written wrongly to the MessageStats column family.
+ *
+ * <p>Correction: {@link ZbColumnFamilies#DEPRECATED_DMN_DECISION_KEY_BY_DECISION_ID_AND_VERSION} ->
+ * {@link ZbColumnFamilies#MESSAGE_STATS}
+ */
+@SuppressWarnings("deprecation") // deals with deprecated column families
+public final class ColumnFamily48Corrector {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DbMigratorImpl.class.getPackageName());
+
+  private static final ZbColumnFamilies CF_UNDER_RECOVERY =
+      ZbColumnFamilies.DEPRECATED_DMN_DECISION_KEY_BY_DECISION_ID_AND_VERSION;
+  private static final ZbColumnFamilies CF_POSSIBLE_TARGET = ZbColumnFamilies.MESSAGE_STATS;
+
+  private final ColumnFamily<DbBytes, DbBytes> recoveryDmnDecisionKeyByDecisionIdAndVersion;
+
+  private final DbString decisionId;
+  private final DbInt decisionVersion;
+
+  private final ColumnFamily<DbString, DbLong> messageStatsColumnFamily;
+  private final DbString messagesDeadlineCountKey;
+  private final DbLong messagesDeadlineCount;
+
+  public ColumnFamily48Corrector(
+      final ZeebeDb<ZbColumnFamilies> zeebeDb, final TransactionContext transactionContext) {
+    decisionId = new DbString();
+    decisionVersion = new DbInt();
+    recoveryDmnDecisionKeyByDecisionIdAndVersion =
+        zeebeDb.createColumnFamily(
+            CF_UNDER_RECOVERY, transactionContext, new DbBytes(), new DbBytes());
+
+    messagesDeadlineCountKey = new DbString();
+    messagesDeadlineCount = new DbLong();
+    messageStatsColumnFamily =
+        zeebeDb.createColumnFamily(
+            CF_POSSIBLE_TARGET,
+            transactionContext,
+            messagesDeadlineCountKey,
+            messagesDeadlineCount);
+  }
+
+  public void correctColumnFamilyPrefix() {
+    recoveryDmnDecisionKeyByDecisionIdAndVersion.forEach(
```

Comment:
Iterating over all key value pairs will lead to OOM if you have more state than memory, esp. since all the pairs end up allocated twice. I have a pending PR which solves this issue, but it's not merged yet.

See https://github.com/camunda/zeebe/pull/16090, specifically the poorly named class MemoryBoundedColumnIteration and how it's used. Obviously the drain method won't work here since you don't want to delete all existing pairs, but we can adapt it.

We could merge your PR first, then mine, then fix it in a follow up as long as we don't release in between

Reply:
Do you think this is blocking a patch release for 8.4?

Reply:
It will cause customers who have to migrate a lot of data here to go out of memory. What do you think the likelihood of having lots of data here, or at least more than fits in memory, is? I guess it's a bit different since we're not moving _everything_ (as opposed to the other issue), but it could still be a lot?

Reply:
Do you want to sync? I need a reviewer on our PR (Berkay's and mine) anyway :upside_down_face:

Reply:
The message stats contains a single entry, and the migration state contains a maximum of 20 entries.

But, there can be many entries in these CFs:
- one for each process instance
- one for each signal subscription
- one for each decision requirements
- one for each decision (there can be multiple per decision requirements)

So, let's fix this.

How does the MemoryBoundedColumnIteration ensure that it is continued once the memory is full?

Reply:
After discussing this in more detail, we decided against adding the ability to chunk these migrations when going out of memory. This is because it's highly unlikely that this migration task will go out of memory.

The memory is increased per operation (i.e. PUT, DELETE) in the transaction. The increase is based on the number of bytes in these operations:
- only the key for DELETE
- the key + the value for PUT

The affected column families may contain many entries, but only a few must be moved/deleted. A move would be a PUT **and** a DELETE. The following entries would need to be moved/deleted:
- In 48, only one entry can be moved to `MESSAGE_STATS`
- In 49, one entry could be moved per active process instance to `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY`
- In 50, only a maximum of 20 entries can be moved to `MIGRATIONS_STATE`

So, only 49 has some actual number of entries to move. Each move is a DELETE of the key and a PUT of the key and value. The key combines the process definition key and the process instance key (twice one long of 8 bytes, so 16 bytes total). The value is a NilValue which is represented by a single byte.

So, a DELETE is just 16 bytes, and a PUT is just 17 bytes. Making each move a total of 33 bytes. At 1 million process instances, this would only take 33MB.